### ISSUE TITLE
sync am (sam-l and sam-u) has a MAJOR bug (kill mchf function)

### ISSUE THREAD

**Issue by DG9BFC** [df8oe/mchf-github#833]
i can stop mchf working with procedure as follows:
switch to sync am
with long press switch to sam-l (or with another to sam-u
then short press and mchf is stopped completely (no audio, no display change, no power button works so you have to kill power supply (pull the plug)
...
if after long press and switch to sam-l i do another long press ... and another ... then no problem with pressing short
but when in sam-l or in sam-u a short press kills mchf
when in sam or am ... no problem
only in sam-l and -u the problem exists
(and i did not activate noise reduction or any other cpu intensive thingy)

**Comment by DD4WH** [comment]
Hi Sigi,
can´t reproduce that here. Which version of the firmware are you using?
Is it a HAL version or the regular version and which date?
73 de Frank

**Comment by df8oe** [comment]
It is only reproducable on HAL version.

**Comment by db4ple** [comment]
It only happens if you switch to FM, if you disable FM it does not happen. Will look into the interaction between SAM & AM and FM Demod and Mode Switching.

**Comment by DG9BFC** [comment]
i reflashed the versin from 5th of march (pre hal version) ... there the problem does NOT exist
 ... when using pre hal version all is well
when in sam-l or sam-u ... a short press of mode button switches to fm
when i use the hal version it has the problem described above
i will use the older version till you have found that bug ... then i can switch over to hal version
disable of fm mode is no solution for me cause i use fm on 10m repeater (dm0max)

**Comment by db4ple** [comment]
Hi,
with my modified HAL source code I cannot reproduce the error anymore (before I could). I reorganized the switching very little but it seems to have done the job. Not sure that I fixed the issue in general (since I don't know what fixed it, there was nothing really suspiscious).

Have to trace it a little more. It is for sure not strictly HAL related, I suppose it is related to stack/memory usage.

**Comment by DG9BFC** [comment]
can you send me the bin for trying it out??
or send a commit and andreas can make a new dayly snapshot??

**Comment by db4ple** [comment]
Hi Sigi,
now I am feeling better, I can reproduce the issue even with my new changes (which I could not before as mentionend). If I switch magnification levels,  the problem is there. It is not the magnification level, it is the switching. Also if I start in magnification x4 I can reproduce the problem. More investigation necessary.

**Comment by db4ple** [comment]
Hi,
I think I found the culprit. When switching away from SAM it could happen that we killed one of the filters used in SAM, which in turn lead to an infinite loop due to too long interrupts or something like this.
Since SAM-L&SAM-U uses a little more horse power it seems that under certain conditions it happens.
Could have happend also with non-HAL software, we were just lucky there. 
I think we should rethink the DEMOD mode switching approach. Possibly we disable the entire processing chain since it is not safe to switch filters then. Of course we could guard all filter uses, but I think the switch over time is about a single frame or two, most likely we won't notice the interruption anyway, and it would make things so much easier to coordinate.

Will push my commit soon for verification.
